processor/memorylimiter/memorylimiter.go

```diff
@@ -238,44 +239,89 @@ func (ml *memoryLimiter) forcingDrop() bool {
 	return atomic.LoadInt64(&ml.forceDrop) != 0
 }
 
-func (ml *memoryLimiter) memCheck() {
-	ms := ml.readMemStats()
-	ml.memLimiting(ms)
+func (ml *memoryLimiter) setForcingDrop(b bool) {
+	var i int64
+	if b {
+		i = 1
+	} else {
+		i = 0
+	}
+	atomic.StoreInt64(&ml.forceDrop, i)
 }
 
-func (ml *memoryLimiter) memLimiting(ms *runtime.MemStats) {
-	if !ml.decision.shouldDrop(ms) {
-		atomic.StoreInt64(&ml.forceDrop, 0)
-	} else {
-		atomic.StoreInt64(&ml.forceDrop, 1)
-		// Force a GC at this point and see if this is enough to get to
-		// the desired level.
+func bytesToMib(ms *runtime.MemStats) zap.Field {
+	return zap.Uint64("cur_mem_mib", ms.Alloc/1024/1024)
+}
+
+func (ml *memoryLimiter) checkMemLimits() {
+	ms := ml.readMemStats()
+
+	ml.logger.Debug("Currently used memory.", bytesToMib(ms))
+
+	if ml.usageChecker.aboveHardLimit(ms) {
+		// Memory usage is above hard limit. Force GC.
+		ml.logger.Warn("Memory usage is above hard limit, force GC.", bytesToMib(ms))
 		runtime.GC()
+		ms = ml.readMemStats()
+		ml.logger.Info("Memory usage after GC.", bytesToMib(ms))
+	}
+
+	if ml.forcingDrop() {
```

Comment:
You can simplify this if your `setForcingDrop` compares the value before setting it, like this:

```go
forceDrop := ml.usageChecker.aboveSoftLimit(ms)
if ml.forcingDrop() && !forceDrop {
  // it would be a great to have this as a gauge!
  ml.logger.Info("Memory usage back within limits, resuming normal operation.", bytesToMib(ms))
}
if !ml.forcingDrop() && forceDrop {
  ml.logger.Warn("Memory usage is above soft limit, begin dropping data.", bytesToMib(ms))
}
ml.setForcingDrop(forceDrop)
```

Reply:
I would do the set of the force dropping before logging, in case we are under pressure we want to start dropping before allocating more memory for logging.

Reply:
Simplified the code as suggested.